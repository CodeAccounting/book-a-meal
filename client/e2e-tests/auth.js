module.exports = {
  'Authentication: Login': (client) => {
    client
      .url('http://localhost:8000')
      .waitForElementVisible('body', 5000)
      .assert.visible('button')
      .click('button')
      .pause(1000)
      .assert.containsText('div.title h2', 'Welcome Back')
      .assert.visible('input[name=email]')
      .assert.visible('input[name=password]')
      .assert.visible('p a[href="/signup?role=customer"]')
      .assert.visible('p a[href="/signup?role=caterer"]')
      .setValue('input[name=email]', 'olisa@emodi.com')
      .setValue('input[name=password]', 'olisaemodi')
      .click('button')
      .pause(3000)
      .assert.containsText('div.navlinks h3', 'Welcome, Olisa')
      .end();
  },

  'Authentication: Customer Signup': (client) => {
    client
      .url('http://localhost:8000')
      .waitForElementVisible('body', 5000)
      .assert.visible('button')
      .click('button')
      .pause(1000)
      .assert.visible('p a[href="/signup?role=customer"]')
      .click('p a[href="/signup?role=customer"]')
      .pause(1000)
      .assert.containsText('div.title h2', 'Start Filling Your Belly')
      .assert.visible('input[name=firstname]')
      .assert.visible('input[name=lastname]')
      .assert.visible('input[name=email]')
      .assert.visible('input[name=password]')
      .assert.visible('input[name=passwordConfirm]')
      .setValue('input[name=firstname]', 'Ola')
      .setValue('input[name=lastname]', 'Shaguy')
      .setValue('input[name=email]', 'ola@shaguy.com')
      .setValue('input[name=password]', 'olashaguy')
      .setValue('input[name=passwordConfirm]', 'olashaguy')
      .click('button')
      .pause(5000)
      .assert.containsText('div.navlinks h3', 'Welcome, Ola')
      .end();
  },

  'Authentication: Caterer Signup': (client) => {
    client
      .url('http://localhost:8000')
      .waitForElementVisible('body', 5000)
      .assert.visible('button')
      .click('button')
      .pause(1000)
      .assert.visible('p a[href="/signup?role=caterer"]')
      .click('p a[href="/signup?role=caterer"]')
      .pause(1000)
      .assert.containsText('div.title h2', 'Start Serving Customers')
      .assert.visible('input[name=businessName]')
      .assert.visible('input[name=email]')
      .assert.visible('input[name=password]')
      .assert.visible('input[name=passwordConfirm]')
      .assert.visible('input[name=phoneNo]')
      .assert.visible('input[name=address]')
      .setValue('input[name=businessName]', 'Lalayum')
      .setValue('input[name=email]', 'lala@yum.com')
      .setValue('input[name=password]', 'lalayumyum')
      .setValue('input[name=passwordConfirm]', 'lalayumyum')
      .setValue('input[name=phoneNo]', '08123456789')
      .setValue('input[name=address]', '3, Church Street, Yaba')
      .click('button')
      .pause(5000)
      .assert.containsText('div.username-circle p', 'L')
      .end();
  }
};
