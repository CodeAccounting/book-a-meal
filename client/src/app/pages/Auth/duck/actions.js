import {
  AUTHENTICATING,
  CLEAR_AUTH_ERROR,
  SIGNIN_SUCCESS,
  SIGNIN_ERROR,
} from './types';

const authenticating = () => ({
  type: AUTHENTICATING
});

const signinSuccess = data => ({
  type: SIGNIN_SUCCESS,
  payload: data
});

const signinFailure = error => ({
  type: SIGNIN_ERROR,
  payload: error
});

const clearAuthError = () => ({
  type: CLEAR_AUTH_ERROR
});

export default {
  signinSuccess,
  signinFailure,
  authenticating,
  clearAuthError,
};
