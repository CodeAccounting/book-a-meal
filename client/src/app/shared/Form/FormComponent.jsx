import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import SigninForm from './SigninForm';
import CustomerSignupForm from './CustomerSignupForm';
import CatererSignupForm from './CatererSignupForm';
import MiniPreLoader from '../Preloader/MiniPreloader';
import { formHelpers, getTouchedFields, formErrorCount } from '../../../helpers';
import { arrayToObject } from '../../../utils';
import { syncValidate, validateRequiredFields } from '../../../helpers/validations';

/**
 * @exports
 * @class FormComponent
 * @extends Component
 * @returns {component} Form
 */
class FormComponent extends Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    submitting: PropTypes.bool.isRequired,
    submitError: PropTypes.string,
    type: PropTypes.string.isRequired,
    meta: PropTypes.shape({
      btnText: PropTypes.string,
      extra: PropTypes.element
    }).isRequired
  };

  static defaultProps = {
    submitError: null
  }

  /**
   * @constructor
   * @memberof Form
   * @param {object} props - props
   * @returns {nothing} Return nothing
   */
  constructor(props) {
    super(props);

    const { type } = props;
    const { formFields } = formHelpers;
    const fields = formFields[type];
    const values = arrayToObject(fields, '');

    if (type === 'catererSignup') values.role = 'caterer';
    if (type === 'customerSignup') values.role = 'customer';

    this.state = {
      type,
      values,
      touched: arrayToObject(fields, false),
      error: arrayToObject(fields, null),
      pristine: true,
      formValid: false,
      asyncValidating: false
    };
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  componentWillMount = () => {
    this.clearFormErrors();
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  clearFormErrors = () => {
    const { type } = this.state;
    const { clearFormError } = formHelpers;

    this.props.dispatch(clearFormError[type]);
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleChange = (event) => {
    const { target } = event;
    const { value, name } = target;

    this.setState(prevState => ({
      values: { ...prevState.values, [name]: value },
      touched: { ...prevState.touched, [name]: true },
      pristine: false,
    }), this.validateForm);
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleFocus = (event) => {
    const { target } = event;
    const { name } = target;

    this.setState(prevState => ({
      error: { ...prevState.error, [name]: null }
    }), () => this.clearFormErrors());
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleBlur = (event) => {
    const { target } = event;
    const { name } = target;

    this.setState(prevState => ({
      touched: { ...prevState.touched, [name]: true },
    }), () => { this.validateField(name); });
  }

  /**
   * @memberof Form
   * @param {string} name
   * @param {string} value
   * @returns {nothing} Returns nothing
   */
  validateField = (name) => {
    const { type } = this.state;
    const error = syncValidate(type)(name, this.state.values);
    const errorValue = error || null;

    this.setState(prevState => ({
      error: { ...prevState.error, [name]: errorValue }
    }), this.validateForm);
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  validateForm = () => {
    const formErrorArrayLength = formErrorCount(this.state.error);
    const touchedFields = getTouchedFields(this.state.touched);
    const { requiredFormFields } = formHelpers;
    const { type } = this.props;
    const requiredFields = requiredFormFields[type];

    if (formErrorArrayLength) {
      this.setState({
        formValid: false
      });
    } else if (!formErrorArrayLength &&
      validateRequiredFields(touchedFields, requiredFields, this.state.values)) {
      this.setState({
        formValid: true
      });
    }
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  submitter = () => {
    const { type } = this.state;
    const { formSubmitMapper } = formHelpers;
    const { values } = this.state;

    return formSubmitMapper[type](values);
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleSubmit = (event) => {
    event.preventDefault();

    return this.props.dispatch(this.submitter());
  }

  /**
   * @memberof Form
   * @returns {JSX} Form
   */
  renderForm = () => {
    const { type } = this.props;
    const formState = { ...this.state };

    const handlers = {
      handleChange: this.handleChange,
      handleBlur: this.handleBlur,
      handleFocus: this.handleFocus,
      handleSubmit: this.handleSubmit
    };

    switch (type) {
      case 'customerSignup':
        return <CustomerSignupForm type={type} state={formState} handlers={handlers} />;
      case 'catererSignup':
        return <CatererSignupForm type={type} state={formState} handlers={handlers} />;
      default:
        return <SigninForm type={type} state={formState} handlers={handlers} />;
    }
  }

  /**
   * @memberof Form
   * @returns {component} Form
   */
  render() {
    const { pristine, formValid } = this.state;
    const { submitting, submitError, meta: { btnText, extra } } = this.props;
    const requiredTextArray = ['catererSignup', 'customerSignup'];

    return (
      <div>
        {submitting &&
          <div className="modal-preloader text-center"><MiniPreLoader /></div>}
        {!submitting &&
          <Fragment>
            {requiredTextArray.includes(this.props.type) &&
            <p className="text-muted mx-auto text-center">
                Fields marked
              <span className="danger"> *</span> are required
            </p>}
            <form onSubmit={this.handleSubmit}>
              {submitError && (
              <p className="danger text-center mb-0">{submitError}</p>
              )}
              {this.renderForm()}
              <button
                className="btn btn-pri btn-block"
                disabled={!formValid || !!pristine || !!submitting}
              >
                {btnText}
              </button>
            </form>
            {extra && extra}
          </Fragment>}
      </div>);
  }
}

export default FormComponent;
