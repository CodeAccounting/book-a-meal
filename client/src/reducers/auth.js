import {
  AUTHENTICATED,
  AUTHENTICATING,
  UNAUTHENTICATED,
  SIGNIN_SUCCESS,
  SIGNIN_ERROR,
  SIGNUP_SUCCESS,
  SIGNUP_ERROR,
  ADD_ORDER_SUCCESS,
  EDIT_ORDER_SUCCESS,
  CLEAR_AUTH_ERROR,
  AUTHENTICATION_ERROR,
} from '../actions/actionTypes';

const initialState = {
  isAuthenticated: !!localStorage.getItem('jwtToken'),
  error: null,
  user: {},
  loading: false
};

/**
 * Authentication Reducer
 * @param {object} state defaults to initalState
 * @param {string} action action object (type, payload)
 * @returns {object} new state
 */
export default (state = initialState, action) => {
  switch (action.type) {
    case AUTHENTICATING:
      return { ...state, loading: true };
    case AUTHENTICATED:
    case SIGNIN_SUCCESS:
    case SIGNUP_SUCCESS:
      return {
        ...state,
        isAuthenticated: true,
        user: action.payload,
        error: null,
        loading: false
      };
    case UNAUTHENTICATED:
    case CLEAR_AUTH_ERROR:
      return {
        ...state,
        isAuthenticated: false,
        error: null,
        loading: false
      };
    case AUTHENTICATION_ERROR:
    case SIGNIN_ERROR:
    case SIGNUP_ERROR:
      return {
        ...state,
        isAuthenticated: false,
        user: {},
        error: action.payload,
        loading: false
      };
    case ADD_ORDER_SUCCESS:
    case EDIT_ORDER_SUCCESS:
      return {
        ...state,
        user: {
          ...state.user,
          address: action.payload.deliveryAddress,
          phoneNo: action.payload.deliveryPhoneNo
        }
      };
    default:
      return state;
  }
};
