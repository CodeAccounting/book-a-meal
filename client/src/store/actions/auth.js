import {
  AUTHENTICATED,
  UNAUTHENTICATED,
  AUTHENTICATION_ERROR,
  AUTHENTICATING,
  CLEAR_AUTH_ERROR,
  SIGNIN_SUCCESS,
  SIGNIN_ERROR,
  SIGNUP_SUCCESS,
  SIGNUP_ERROR,
} from '../types';

const authenticating = () => ({
  type: AUTHENTICATING
});

const signinSuccess = data => ({
  type: SIGNIN_SUCCESS,
  payload: data
});

const signinFailure = error => ({
  type: SIGNIN_ERROR,
  payload: error
});

const signupSuccess = data => ({
  type: SIGNUP_SUCCESS,
  payload: data
});

const signupFailure = error => ({
  type: SIGNUP_ERROR,
  payload: error
});

const authenticationSuccess = data => ({
  type: AUTHENTICATED,
  payload: data
});

const authenticationFailure = error => ({
  type: AUTHENTICATION_ERROR,
  payload: error
});

const clearAuthError = () => ({
  type: CLEAR_AUTH_ERROR
});

const resetUser = () => ({
  type: UNAUTHENTICATED
});

export default {
  resetUser,
  signinSuccess,
  signinFailure,
  authenticating,
  clearAuthError,
  signupSuccess,
  signupFailure,
  authenticationSuccess,
  authenticationFailure,
};
