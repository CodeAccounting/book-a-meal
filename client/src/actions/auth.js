import instance from '../config/axios';
import { errorHandler } from '../utils';
import { auth as authAPI } from '../utils/api/auth';
import {
  AUTHENTICATED,
  UNAUTHENTICATED,
  AUTHENTICATION_ERROR,
  AUTHENTICATING,
  CLEAR_AUTH_ERROR,
  SIGNIN_SUCCESS,
  SIGNIN_ERROR,
  SIGNUP_SUCCESS,
  SIGNUP_ERROR,
} from './actionTypes';

export const authenticating = () => ({
  type: AUTHENTICATING
});

export const signinSuccess = data => ({
  type: SIGNIN_SUCCESS,
  payload: data
});

export const signinFailure = error => ({
  type: SIGNIN_ERROR,
  payload: error
});

export const signupSuccess = data => ({
  type: SIGNUP_SUCCESS,
  payload: data
});

export const signupFailure = error => ({
  type: SIGNUP_ERROR,
  payload: error
});

export const authenticationSuccess = data => ({
  type: AUTHENTICATED,
  payload: data
});

export const authenticationFailure = error => ({
  type: AUTHENTICATION_ERROR,
  payload: error
});

export const clearAuthError = () => ({
  type: CLEAR_AUTH_ERROR
});

export const resetUser = () => ({
  type: UNAUTHENTICATED
});

/**
 * Authenticates User
 * @function auth
 * @param {string} type
 * @param {object} user
 * @param {func} dispatch
 * @returns {void}
 */
export const auth = type => user => async (dispatch) => {
  try {
    dispatch(authenticating());

    const response = await authAPI(type)(user);

    localStorage.setItem('jwtToken', response.data.token);

    const dispatchType = type === 'signin' ? signinSuccess : signupSuccess;

    dispatch(dispatchType(response.data.user));
  } catch (error) {
    const errorResponse = errorHandler(error);

    const dispatchType = type === 'signin' ? signinFailure : signupFailure;

    dispatch(dispatchType(errorResponse.response));
  }
};

/**
 * Refreshe User Token
 * @function authenticateUser
 * @param {func} dispatch
 * @returns {void}
 */
export const authenticateUser = () => async (dispatch) => {
  try {
    dispatch(authenticating());

    const res = await instance.get('/auth/refresh_token');

    localStorage.setItem('jwtToken', res.data.token);

    dispatch(authenticationSuccess(res.data.user));
  } catch (error) {
    const errorResponse = errorHandler(error);

    localStorage.removeItem('jwtToken');

    dispatch(authenticationFailure(errorResponse.response));
  }
};

/**
 * Logs user out
 * @function logout
 * @param {func} dispatch
 * @returns {void}
 */
export const logout = () => (dispatch) => {
  localStorage.removeItem('jwtToken');
  dispatch(resetUser());
};
