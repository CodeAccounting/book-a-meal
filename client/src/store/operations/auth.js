import instance from '../../config/axios';
import { errorHandler } from '../../utils';
import { auth as authAPI } from '../../services/api/auth';
import {
  resetUser,
  signinSuccess,
  signinFailure,
  signupSuccess,
  signupFailure,
  authenticating,
  clearAuthError,
  authenticationSuccess,
  authenticationFailure,
} from '../actions/auth';

const auth = type => user => async (dispatch) => {
  try {
    dispatch(authenticating());

    const response = await authAPI(type)(user);

    localStorage.setItem('jwtToken', response.data.token);

    // if type is signin dispatch signinSuccess else dispatch signupSuccess
    const dispatchType = type === 'signin' ? signinSuccess : signupSuccess;
    dispatch(dispatchType(response.data.user));
  } catch (error) {
    const errorResponse = errorHandler(error);

    // if type is signin dispatch signinFailure else dispatch signupFailure
    const dispatchType = type === 'signin' ? signinFailure : signupFailure;
    dispatch(dispatchType(errorResponse.response));
  }
};

const authenticateUser = () => async (dispatch) => {
  try {
    dispatch(authenticating());

    const res = await instance.get('/auth/refresh_token');

    localStorage.setItem('jwtToken', res.data.token);

    dispatch(authenticationSuccess(res.data.user));
  } catch (error) {
    const errorResponse = errorHandler(error);

    localStorage.removeItem('jwtToken');

    dispatch(authenticationFailure(errorResponse.response));
  }
};

const logout = () => (dispatch) => {
  localStorage.removeItem('jwtToken');
  dispatch(resetUser());
};

export default {
  auth,
  logout,
  resetUser,
  signinSuccess,
  signinFailure,
  signupSuccess,
  signupFailure,
  authenticating,
  clearAuthError,
  authenticateUser,
  authenticationSuccess,
  authenticationFailure,
};
