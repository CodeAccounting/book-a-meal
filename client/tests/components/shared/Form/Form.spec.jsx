import React from 'react';
import FormComponent from '../../../../src/components/shared/Form';
import { clearAuthError } from '../../../../src/actions/auth';
import { mainFormSetup as setup, formComponentSetup } from '../../../../tests/setup/formSetup';

const signinState = {
  type: 'signin',
  values: {
    email: '',
    password: ''
  },
  touched: { email: false, password: false },
  error: { email: null, password: null },
  pristine: true,
  formValid: false,
  asyncValidating: false
};

const meta = {
  btnText: 'SIGN IN',
  extra: <p>something</p>
};

describe('Form', () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { shallowRoot } = setup('signin', meta);

    expect(toJson(shallowRoot)).toMatchSnapshot();
  });

  it('renders customer\'s signup form correctly', () => {
    const { shallowRoot } = setup('customerSignup', meta);

    expect(toJson(shallowRoot)).toMatchSnapshot();
  });

  it('renders caterer\'s signup form correctly', () => {
    const { shallowRoot } = setup('catererSignup', meta);

    expect(toJson(shallowRoot)).toMatchSnapshot();
  });

  it('renders addMeal form correctly', () => {
    const { shallowRoot } = setup('addMeal', meta);

    expect(toJson(shallowRoot)).toMatchSnapshot();
  });

  it('renders editMeal form correctly', () => {
    const { shallowRoot } = setup('editMeal', meta);

    expect(toJson(shallowRoot)).toMatchSnapshot();
  });

  it('disables submit button when form is clean', () => {
    const { shallowRoot } = setup('signin', meta);

    expect(shallowRoot.find('button[disabled=true]')).toBeTruthy();
  });

  it('disables submit button when submitting form', () => {
    const { props, dispatchMock } = setup('signin', meta);
    const newProps = {
      ...props,
      submitting: true
    };
    const shallowRoot = shallow(<FormComponent {...newProps} dispatch={dispatchMock} />);

    expect(toJson(shallowRoot)).toMatchSnapshot();
    expect(shallowRoot.find('button[disabled=true]')).toBeTruthy();
  });

  it('shows error alert and disables submit button when there\'s a submit error', () => {
    const { props, dispatchMock } = setup('signin', meta);
    const newProps = {
      ...props,
      submitError: 'Username/Password do not match'
    };

    const shallowRoot = shallow(<FormComponent {...newProps} dispatch={dispatchMock} />);

    expect(toJson(shallowRoot)).toMatchSnapshot();
    expect(shallowRoot.find('p.danger')).toBeTruthy();
    expect(shallowRoot.find('button[disabled=true]')).toBeTruthy();
  });

  it('changes form state when prop: type changes', () => {
    const { props, dispatchMock } = setup('signin', meta);

    const shallowRoot = shallow(<FormComponent {...props} dispatch={dispatchMock} />);

    shallowRoot.setProps({
      type: 'editMeal'
    });

    expect(shallowRoot.state().type).toEqual('editMeal');
  });

  it('does not change form state when prop: type does not change but props update', () => {
    const { props, dispatchMock } = setup('signin', meta);

    const shallowRoot = shallow(<FormComponent {...props} dispatch={dispatchMock} />);

    shallowRoot.setProps({
      type: 'signin',
      test: 'test'
    });

    expect(shallowRoot.state().type).toEqual('signin');
    expect(shallowRoot.state().test).toEqual(undefined);
  });

  describe('test for right input', () => {
    it('calls handleChange on input change for email field', () => {
      const { mountRoot, dispatchMock } = setup('signin', meta);
      const wrapper = mountRoot.find(FormComponent);

      const changeState = {
        ...signinState,
        values: { ...signinState.values, email: 'iverenshaguy@gmail.com' },
        touched: { ...signinState.touched, email: true },
        error: { email: null, password: null },
        pristine: false
      };

      const event = { target: { name: 'email', value: 'iverenshaguy@gmail.com' } };

      wrapper.find('input[name="email"]').simulate('focus');
      expect(dispatchMock).toHaveBeenCalledWith(clearAuthError());

      wrapper.find('input[name="email"]').simulate('change', event);
      expect(wrapper.instance().state).toEqual(changeState);

      wrapper.find('input[name="email"]').simulate('blur');
      expect(wrapper.instance().state).toEqual(changeState);
    });

    it('calls handleChange on checkbox change for vegetarian field', () => {
      const { state } = formComponentSetup('addMeal');
      const { mountRoot } = setup('addMeal', meta);
      const wrapper = mountRoot.find(FormComponent);

      const changeState = {
        ...state,
        type: 'addMeal',
        values: { ...state.values, vegetarian: 'true', imageUrl: 'http://res.cloudinary.com/iverenshaguy/image/upload/v1532540264/bookameal/default-img.jpg' },
        touched: { ...state.touched, vegetarian: true },
        pristine: false
      };

      const event = { target: { name: 'vegetarian', type: 'checkbox', checked: 'true' } };

      wrapper.find('input[name="vegetarian"]').simulate('change', event);
      expect(wrapper.instance().state).toEqual(changeState);
    });

    it('submits valid form', () => {
      const { mountRoot, dispatchMock } = setup('signin', meta);
      const wrapper = mountRoot.find(FormComponent);

      const newState = {
        ...signinState,
        values: { email: 'iverenshaguy@gmail.com', password: 'iverenshaguy' },
        touched: { email: true, password: true },
        error: { email: null, password: null },
        pristine: false,
        formValid: true
      };

      const emailEvent = { target: { name: 'email', value: 'iverenshaguy@gmail.com' } };
      const passwordEvent = { target: { name: 'password', value: 'iverenshaguy' } };

      wrapper.find('input[name="email"]').simulate('focus');
      wrapper.find('input[name="email"]').simulate('change', emailEvent);
      wrapper.find('input[name="password"]').simulate('focus');
      wrapper.find('input[name="password"]').simulate('change', passwordEvent);

      expect(wrapper.instance().state).toEqual(newState);

      wrapper.find('form').simulate('submit', { preventDefault() { } });
      expect(dispatchMock).toHaveBeenCalled();
    });

    it('submits valid editMeal form', () => {
      const { mountRoot, dispatchMock } = setup('editMeal', meta);
      const wrapper = mountRoot.find(FormComponent);

      const titleEvent = { target: { name: 'title', value: 'Rice' } };

      wrapper.find('input[name="title"]').simulate('focus');
      wrapper.find('input[name="title"]').simulate('change', titleEvent);

      wrapper.find('form').simulate('submit', { preventDefault() { } });
      expect(dispatchMock).toHaveBeenCalled();
    });

    it('calls dispatch on file input change', () => {
      const { mountRoot, dispatchMock } = setup('editMeal', meta);
      const wrapper = mountRoot.find(FormComponent);

      const event = {
        target: {
          files: [new Blob([''], {
            type: 'image/jpeg',
            size: '500'
          })]
        }
      };

      mocksdk.storage().refFromURL = () => ({
        delete: jest.fn()
      });

      mocksdk.storage().ref = () => ({
        put: () => ({
          ref: {
            getDownloadURL: () => ({
              then: (fn) => {
                fn('http.url.test');
              }
            })
          }
        })
      });

      wrapper.find('input[name="imageUrl"]').simulate('change', event);

      expect(dispatchMock).toHaveBeenCalled();
    });
  });

  describe('test for wrong input', () => {
    it('sync validates field and form on input change and blur', () => {
      const { state } = formComponentSetup('customerSignup');
      const { mountRoot } = setup('customerSignup', meta);
      const wrapper = mountRoot.find(FormComponent);


      const changeState = {
        ...state,
        values: { ...state.values, email: 'emiolaolasanmi', role: 'customer' },
        touched: { ...state.touched, email: true },
        error: { ...state.error, email: 'Invalid email address!', password: null },
        pristine: false,
        type: 'customerSignup',
      };

      const blurState = {
        ...changeState,
        asyncValidating: false
      };

      const event = { target: { name: 'email', value: 'emiolaolasanmi' } };

      wrapper.find('input[name="email"]').simulate('focus');
      wrapper.find('input[name="email"]').simulate('blur');
      expect(wrapper.instance().state).toEqual(({
        ...changeState,
        error: {
          ...changeState.error, email: 'Required!', password: null
        },
        pristine: true,
        values: { ...state.values, role: 'customer' }
      }));

      wrapper.find('input[name="email"]').simulate('change', event);
      expect(wrapper.instance().state).toEqual({ ...changeState, error: { ...changeState.error, email: 'Required!' } });

      wrapper.find('input[name="email"]').simulate('blur', event);
      expect(wrapper.instance().state).toEqual(blurState);

      wrapper.find('input[name="password"]').simulate('focus');
      wrapper.find('input[name="password"]').simulate('blur');
      expect(wrapper.instance().state).toEqual(({
        ...changeState,
        error: { ...changeState.error, password: 'Required!' },
        touched: { ...changeState.touched, email: true, password: true },
        values: { ...changeState.values, email: 'emiolaolasanmi', password: '' }
      }));
    });
  });
});
