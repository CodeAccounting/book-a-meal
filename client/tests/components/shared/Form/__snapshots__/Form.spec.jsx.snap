// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form disables submit button when submitting form 1`] = `
<div>
  <div
    className="modal-preloader text-center"
  >
    <MiniPreloader />
  </div>
</div>
`;

exports[`Form renders addMeal form correctly 1`] = `
<div>
  <React.Fragment>
    <p
      className="text-muted mx-auto text-center"
    >
      Fields marked
      <span
        className="danger"
      >
         *
      </span>
       are required
    </p>
    <form
      onSubmit={[Function]}
    >
      <MealForm
        handlers={
          Object {
            "handleBlur": [Function],
            "handleChange": [Function],
            "handleFocus": [Function],
            "handleSubmit": [Function],
          }
        }
        meal={null}
        state={
          Object {
            "asyncValidating": false,
            "error": Object {
              "description": null,
              "imageUrl": null,
              "price": null,
              "title": null,
              "vegetarian": null,
            },
            "formValid": false,
            "pristine": true,
            "touched": Object {
              "description": false,
              "imageUrl": false,
              "price": false,
              "title": false,
              "vegetarian": false,
            },
            "type": "addMeal",
            "values": Object {
              "description": "",
              "imageUrl": "http://res.cloudinary.com/iverenshaguy/image/upload/v1532540264/bookameal/default-img.jpg",
              "price": "",
              "title": "",
              "vegetarian": false,
            },
          }
        }
        type="addMeal"
        uploading={null}
      />
      <button
        className="btn btn-pri btn-block"
        disabled={true}
      >
        SIGN IN
      </button>
    </form>
    <p>
      something
    </p>
  </React.Fragment>
</div>
`;

exports[`Form renders caterer's signup form correctly 1`] = `
<div>
  <React.Fragment>
    <p
      className="text-muted mx-auto text-center"
    >
      Fields marked
      <span
        className="danger"
      >
         *
      </span>
       are required
    </p>
    <form
      onSubmit={[Function]}
    >
      <CatererSignupForm
        handlers={
          Object {
            "handleBlur": [Function],
            "handleChange": [Function],
            "handleFocus": [Function],
            "handleSubmit": [Function],
          }
        }
        state={
          Object {
            "asyncValidating": false,
            "error": Object {
              "businessAddress": null,
              "businessName": null,
              "businessPhoneNo": null,
              "email": null,
              "password": null,
              "passwordConfirm": null,
              "role": null,
            },
            "formValid": false,
            "pristine": true,
            "touched": Object {
              "businessAddress": false,
              "businessName": false,
              "businessPhoneNo": false,
              "email": false,
              "password": false,
              "passwordConfirm": false,
              "role": false,
            },
            "type": "catererSignup",
            "values": Object {
              "businessAddress": "",
              "businessName": "",
              "businessPhoneNo": "",
              "email": "",
              "password": "",
              "passwordConfirm": "",
              "role": "caterer",
            },
          }
        }
        type="catererSignup"
      />
      <button
        className="btn btn-pri btn-block"
        disabled={true}
      >
        SIGN IN
      </button>
    </form>
    <p>
      something
    </p>
  </React.Fragment>
</div>
`;

exports[`Form renders correctly 1`] = `
<div>
  <React.Fragment>
    <form
      onSubmit={[Function]}
    >
      <SigninForm
        handlers={
          Object {
            "handleBlur": [Function],
            "handleChange": [Function],
            "handleFocus": [Function],
            "handleSubmit": [Function],
          }
        }
        state={
          Object {
            "asyncValidating": false,
            "error": Object {
              "email": null,
              "password": null,
            },
            "formValid": false,
            "pristine": true,
            "touched": Object {
              "email": false,
              "password": false,
            },
            "type": "signin",
            "values": Object {
              "email": "",
              "password": "",
            },
          }
        }
        type="signin"
      />
      <button
        className="btn btn-pri btn-block"
        disabled={true}
      >
        SIGN IN
      </button>
    </form>
    <p>
      something
    </p>
  </React.Fragment>
</div>
`;

exports[`Form renders customer's signup form correctly 1`] = `
<div>
  <React.Fragment>
    <p
      className="text-muted mx-auto text-center"
    >
      Fields marked
      <span
        className="danger"
      >
         *
      </span>
       are required
    </p>
    <form
      onSubmit={[Function]}
    >
      <CustomerSignupForm
        handlers={
          Object {
            "handleBlur": [Function],
            "handleChange": [Function],
            "handleFocus": [Function],
            "handleSubmit": [Function],
          }
        }
        state={
          Object {
            "asyncValidating": false,
            "error": Object {
              "email": null,
              "firstname": null,
              "lastname": null,
              "password": null,
              "passwordConfirm": null,
              "role": null,
            },
            "formValid": false,
            "pristine": true,
            "touched": Object {
              "email": false,
              "firstname": false,
              "lastname": false,
              "password": false,
              "passwordConfirm": false,
              "role": false,
            },
            "type": "customerSignup",
            "values": Object {
              "email": "",
              "firstname": "",
              "lastname": "",
              "password": "",
              "passwordConfirm": "",
              "role": "customer",
            },
          }
        }
        type="customerSignup"
      />
      <button
        className="btn btn-pri btn-block"
        disabled={true}
      >
        SIGN IN
      </button>
    </form>
    <p>
      something
    </p>
  </React.Fragment>
</div>
`;

exports[`Form renders editMeal form correctly 1`] = `
<div>
  <React.Fragment>
    <p
      className="text-muted mx-auto text-center"
    >
      Fields marked
      <span
        className="danger"
      >
         *
      </span>
       are required
    </p>
    <form
      onSubmit={[Function]}
    >
      <MealForm
        handlers={
          Object {
            "handleBlur": [Function],
            "handleChange": [Function],
            "handleFocus": [Function],
            "handleSubmit": [Function],
          }
        }
        meal={
          Object {
            "description": "",
            "id": "81211c24-51c0-46ec-b1e0-18db55880954",
            "imageUrl": "http://www.preciouscore.com/wp-content/uploads/2017/11/How-to-cook-jollof-rice-in-the-oven-750x500.jpg",
            "price": "2500.00",
            "title": "Jollof Rice and Chicken",
            "vegetarian": false,
          }
        }
        state={
          Object {
            "asyncValidating": false,
            "error": Object {
              "description": null,
              "imageUrl": null,
              "price": null,
              "title": null,
              "vegetarian": null,
            },
            "formValid": true,
            "pristine": true,
            "touched": Object {
              "description": false,
              "imageUrl": false,
              "price": false,
              "title": false,
              "vegetarian": false,
            },
            "type": "editMeal",
            "values": Object {
              "description": "",
              "id": "81211c24-51c0-46ec-b1e0-18db55880954",
              "imageUrl": "http://www.preciouscore.com/wp-content/uploads/2017/11/How-to-cook-jollof-rice-in-the-oven-750x500.jpg",
              "price": "2500.00",
              "title": "Jollof Rice and Chicken",
              "vegetarian": false,
            },
          }
        }
        type="editMeal"
        updating={false}
        uploading={null}
      />
      <button
        className="btn btn-pri btn-block"
        disabled={false}
      >
        SIGN IN
      </button>
    </form>
    <p>
      something
    </p>
  </React.Fragment>
</div>
`;

exports[`Form shows error alert and disables submit button when there's a submit error 1`] = `
<div>
  <React.Fragment>
    <form
      onSubmit={[Function]}
    >
      <p
        className="danger text-center mb-0"
      >
        Username/Password do not match
      </p>
      <SigninForm
        handlers={
          Object {
            "handleBlur": [Function],
            "handleChange": [Function],
            "handleFocus": [Function],
            "handleSubmit": [Function],
          }
        }
        state={
          Object {
            "asyncValidating": false,
            "error": Object {
              "email": null,
              "password": null,
            },
            "formValid": false,
            "pristine": true,
            "touched": Object {
              "email": false,
              "password": false,
            },
            "type": "signin",
            "values": Object {
              "email": "",
              "password": "",
            },
          }
        }
        type="signin"
      />
      <button
        className="btn btn-pri btn-block"
        disabled={true}
      >
        SIGN IN
      </button>
    </form>
    <p>
      something
    </p>
  </React.Fragment>
</div>
`;
